### ISSUE TITLE
twisted.web timeout of 12 hours is too long

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3746]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| ivank reported|
|-|-|
|Trac ID|trac#3746|
|Type|defect|
|Created|2009-04-01 06:58:41Z|
|Branch|https://github.com/twisted/twisted/tree/3746-lukasa-tweb-timeout|

Firefox tries to keep the connection open for 5 minutes; IE 1 or 2 minutes. Right now, Twisted keeps connections open for 12 hours if the browser does not close them. Apache's default timeout is 15 seconds, and nginx's is 65 seconds. A short timeout would not disrupt responses that are just taking a while to generate.

The timeout is in three places:
 - ```twisted/web/http.py:    timeOut = 60 * 60 * 12```
 - ```twisted/web/http.py:    def __init__(self, logPath=None, timeout=60*60*12):```
 - ```twisted/web/server.py:    def __init__(self, resource, logPath=None, timeout=60*60*12):```

Attachments:

* [reduce_http_timeout_3746.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a59/a59b456bed547cea87668d828fdc80f20175c81f/3cda9dba5d50145a288bf0da05ac1ec5d5f67b84.patch) (859 bytes) - added by Rotund on 2010-02-23 21:44:29Z - Reduce default timeout to 15 seconds

<details><summary>Searchable metadata</summary>

```
trac-id__3746 3746
type__defect defect
reporter__ivank ivank
priority__high high
milestone__ 
branch__3746_lukasa_tweb_timeout 3746-lukasa-tweb-timeout
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__None None
time__1238569121000000 1238569121000000
changetime__1621012983029914 1621012983029914
version__None None
owner__glyph glyph
cc__thijs cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_1"></a>|@thijstriemstra commented|
|-|-|

add markup to ticket description

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

The timeout is a public feature of the API.  Applications can set it to anything they like.  I _would_ be interested in adding some disconnect pressure based on how close the process is to its file descriptor limit, but I don't think I'm very interested in playing guess-the-timeout-that-is-appropriate-for-all-applications.

[#11553](https://github.com/twisted/twisted/issues/11553) and [#11015](https://github.com/twisted/twisted/issues/11015) are somewhat related to this, I think.

Unless you have a motivation for timing out connections for a different reason?

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra commented|
|-|-|

Replying to [exarkun](#note_2):
> I _would_ be interested in adding some disconnect pressure based on how close the process is to its file descriptor limit, but I don't think I'm very interested in playing guess-the-timeout-that-is-appropriate-for-all-applications.

That 12 hours is already a guess-the-timeout-that-is-appropriate-for-all-applications, but simply too much :) Why not lower the default to something that's closer to Apache's value? It won't change the fact it's a public feature of the API and applications will still be able to set it. It's just hard to imagine that any Twisted application would use a 12 hour timeout vs the values other webservers use..

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|ivank commented|
|-|-|

My description was incorrect. At least in Firefox, 'network.http.keep-alive.timeout' = 300 is just a hint sent to the webserver (as Keep-Alive: 300). Firefox keeps the connection open until the webserver closes it.

So, by default, a Firefox visitor loading page on unproxied twisted.web website would keep the connection open for 12 hours. I don't think anyone should expect this to happen.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_5"></a>|@jyknight commented|
|-|-|

Agreed, 12 hours is **way** too long. I made web2 have two timeouts: one is 15sec (between requests) and one is 4 minutes (while reading a request). There is no timeout once the request has finished being read until the server finishes writing the response. Those are good default values: twisted.web should copy them.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|ivank commented|
|-|-|

This has security consequences because the default timeout value makes it very easy to intentionally or accidentally DoS a twisted.web server.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|ivank commented|
|-|-|

Lowering the timeout requires implementing a 'noisy' mode for HTTPChannel, or else the log will be filled with messages from:

```python
log.msg("Timing out client: %s" % str(self.transport.getPeer()))
```

The noisy mode might also log connecting and disconnecting clients.

Do 'noisy' messages require tests?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

I'm not sure that log message is worth keeping.

**Comment by twisted-trac** [comment]
|<img alt="Rotund's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|Rotund commented|
|-|-|

Cherokee also defaults to 15 seconds.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|ivank commented|
|-|-|

Another thing we could do in this branch is to convince browsers to close the connection by sending them a `Keep-Alive: timeout=N` header[1].  This would reduce the number of `TIME_WAIT` sockets on the server.

[1] browser behavior varies; see http://wiki.nginx.org/HttpCoreModule#keepalive_timeout

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_13"></a>|@hawkowl commented|
|-|-|

I updated this; please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_14"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Changes look good. 5 minutes is very conservative for my taste :)

 we can argue forever on this default value... but I have another comment: where is the associated test, I don't see any changes in the existing tests ?

Is it ok to merge without test coverage?

Please check my comments and merge, or fix and merge.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_15"></a>|@Lukasa removed owner|
|-|-|

Ok, I'm taking this back up from hawkie.

Like adi, I think that five minutes is too conservative, so I've updated hawkie's patch to do a one minute timeout instead. I've also added a test of that default behaviour. Please re-review.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_16"></a>|@Lukasa commented|
|-|-|

In changeset 396071d3a903649893326a3f061a2b2567ca973c
```
#!CommitTicketReference repository="" revision="396071d3a903649893326a3f061a2b2567ca973c"
Fix the name of the topfile for #3746.
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph set owner to @glyph|
|-|-|

Thanks! Just waiting for the builders.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph set status to `closed`|
|-|-|

In changeset 08795793d26040047d1e0aa5de865f686eedc977
```
#!CommitTicketReference repository="" revision="08795793d26040047d1e0aa5de865f686eedc977"
Merge 3746-lukasa-tweb-timeout: Brief description

Author: hawkowl, lukasa

Reviewer: adiroiban, glyph

Fixes: #3746

twisted.web.http.HTTPFactory now times connections out after one minute of no data from the client being received, before the request is complete, rather than twelve hours.
```

**Comment by twisted-trac** [comment]
|<img alt="BMtthws's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_19"></a>|BMtthws commented|
|-|-|

Can someone confirm if this timeout change will interfere with a long-poll request that waits up to 2 minutes for a response?
